# HG changeset patch
# User ylw <ylw633@gmail.com>
# Date 1429758896 18000
#      Wed Apr 22 22:14:56 2015 -0500
# Branch scite-mql
# Node ID 7bdba7c9c807502c9d5592a1331bdf0485a34dc1
# Parent  57c52893ca01366bd0eec7580c9333d9d6f2b8bd
scite-mql

diff -r 57c52893ca01 -r 7bdba7c9c807 pack/SciTEGlobal.properties
--- a/pack/SciTEGlobal.properties	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/SciTEGlobal.properties	Wed Apr 22 22:14:56 2015 -0500
@@ -32,7 +32,7 @@
 
 # Установка в 1 снижает точность отображения активности/неактивности кнопок на панели инструментов
 # Это может улучшить работу на относительно медленных машинах
-#~ undo.redo.lazy=1
+undo.redo.lazy=1
 
 # В версии для GTK+ определяет начальные размеры окна диалога выбора файлов вызываемого командами Открыть(Open) и Сохранить(Save). В Windows данный параметр не работает.
 #~ fileselector.width=800
@@ -58,7 +58,7 @@
 
 # 1 включает запрос сохранения при сборке (F7) файла. О - сохраняет изменения без запроса.
 #~ are.you.sure.for.build=1
-#~ save.all.for.build=1
+save.all.for.build=1
 
 # 1 закрывает SciTE при закрытии последней вкладки
 #~ quit.on.close.last=1
@@ -82,10 +82,10 @@
 check.if.already.open=1
 
 # Определяет используемый режим языка для создаваемого (Ctrl+N) файла, прежде чем ему будет назначено имя
-default.file.ext=.txt
+default.file.ext=.mq4
 
 # Если заданное в ком. строке имя файла без расширения не найдено, то SciTE будет пытаться открыть одноименный файл с указанными ниже расширениями
-source.default.extensions=.h|.cxx|.bat
+source.default.extensions=.mq4|.mqh
 
 # 1 - диалог "Открыть" открывается в папке текущего файла, 0 - где открывали последний раз
 open.dialog.in.file.directory=1
@@ -100,7 +100,7 @@
 ensure.consistent.line.ends=1
 
 #~ save.deletes.first=1
-
+save.check.modified.time=1
 # При отсутствии перевода вставить знак...
 #~ translation.missing=***
 
@@ -138,12 +138,22 @@
 
 # Подсветка всех вхождений текущего слова в документе (встроенная возможность SciTE)
 # Совет: Лучше использовать скриптовый вариант (см. ниже). В нем и возможностей - больше и работает быстрее.
-#~ highlight.current.word=1
+#highlight.current.word=0
 #~ highlight.current.word.by.style=1
-#~ highlight.current.word.colour=#00D040
+#highlight.current.word.colour=#00D040
+
+[Smart Highlight]
+smart.highlight=1
+smart.highlight.matchcase=1
+# number,style,colour,@alpha
+# number: 0-31
+# style: plain squiggle tt diagonal strike hidden roundbox box
+smart.highlight.style=18,roundbox,#FFD700,@255
+smart.highlight.minlen=1
+smart.highlight.maxnum=100
 
 # Авто подсветка текста, который совпадает с текущим словом или выделением (highlighting_identical_text.lua)
-highlighting.identical.text=0
+#highlighting.identical.text=1
 # Ограничение на max кол-во вхождений (50 - по умолчанию, 0 - без ограничения).
 #~ highlighting.identical.text.max=100
 
@@ -205,7 +215,7 @@
 [Output Screen]      Положение и размер окна консоли
 # Автоматически очищает окно консоли перед выполнением любого скрипта
 # т.е. выполняет команду output:ClearAll()  (LUA)
-#~ clear.before.execute=1
+clear.before.execute=1
 
 # 1 - окно консоли справа, 0 - внизу
 split.vertical=0
@@ -217,7 +227,6 @@
 #~ output.horizontal.scrollbar=0
 #~ output.horizontal.scroll.width=10000
 #~ output.scroll=0
-
 output.cache.layout=3
 
 # 1 - после завершения команды выводить время её выполнения (в секундах)
@@ -284,20 +293,20 @@
 #~ control.char.symbol=°
 
 # Показывать пробелы и табуляции (0/1)
-view.whitespace=1
-view.indentation.whitespace=1
+view.whitespace=0
+view.indentation.whitespace=0
 view.indentation.guides=1
 highlight.indentation.guides=1
 
 [Indent]             Настройки отступов и табуляции
 # Размер отступа
-tabsize=4
+tabsize=3
 
 # Размер отступа, для использования при автоматических отступах
-indent.size=4
+indent.size=3
 
 # При 0 символ табуляции при вводе автоматически заменяется на пробелы
-#~ use.tabs=0
+use.tabs=0
 
 # При 1 значения indent.size и use.tabs берутся из открытого документа
 indent.auto=1
@@ -319,10 +328,10 @@
 edge.column=100
 
 # Способ показа превышения допустимой ширины (0-без показа, 1-линия, 2-цвет)
-edge.mode=1
+edge.mode=0
 
 # Переносить длинные строки (0/1)
-wrap=1
+wrap=0
 
 #~ wrap.style=2
 wrap.visual.flags=3
@@ -376,13 +385,13 @@
 
 [Find and Replace]   Поиск и замена
 # ------------ ПОИСК В ТЕКУЩЕМ ФАЙЛЕ ------------
-#~ find.in.files.close.on.find=0
-#~ find.replace.matchcase=1
-#~ find.replace.escapes=1
-#~ find.replace.regexp=1
+find.in.files.close.on.find=0
+find.replace.matchcase=0
+find.replace.escapes=0
+find.replace.regexp=0
 find.replace.regexp.posix=1
-#~ find.replace.wrap=0
-#~ find.replacewith.focus=0
+find.replace.wrap=1
+find.replacewith.focus=1
 
 # Если 1, то вместо диалогового окна поиска используется панель
 find.use.strip=1
@@ -425,18 +434,7 @@
   #~ find.command=grep --line-number "$(find.what)" $(find.files)
 
 # Список расширений файлов, в которых будет производится поиск
-find.files=\
-*|\
-*.lua *.properties *.api|\
-*.html *.htm *.php *.js|\
-*.properties *.html *.xml *.iface|\
-*.vbs *.js *.wsf *.hta *.lua|\
-make* *.mak *.bat *.cmd *.e|\
-*.vb *.vbs *.bas *.frm *.cls *.ctl|\
-*.java *.js *.py *.pl *.rb *.cgi *.lua *.conf|\
-*.asm *.c *.cc *.cpp *.cxx *.cs *.h *.hh *.hxx *.hpp *.rc|\
-*.pas *.dpr *.dpk *.dfm *.res|\
-*.idl *.odl *.rc *.rc2 *.dlg *.def
+find.files=*
 
 # Параметры скрипта FindText.lua:
 # Поиск с учётом регистра
@@ -463,7 +461,7 @@
 # &f = имя файла; &F = полный путь к файлу
 # &d = дата файла; &D = текущая дата
 # &t = время файла; &T = полное время
-print.header.format=$(FileNameExt) — Распечатано $(CurrentDate), $(CurrentTime) — Страница $(CurrentPage)
+print.header.format=$(FileNameExt) - On $(CurrentDate), $(CurrentTime) — Page $(CurrentPage)
 print.footer.format=$(FilePath) — File date: $(FileDate) — File time: $(FileTime)
 
 # Стиль текста для колонтитулов
@@ -551,15 +549,15 @@
 # Имя файла локализации интерфейса
 #~ locale.properties=locale-ru.properties
 # Кодировка окна редактирования. Изменяется скриптом CodePage.lua. Используется для функций string.{upper|lower|len}
-editor.code.page=1251
+editor.code.page=65001
 
 # Unicode UTF-8
-#~ code.page=65001
+code.page=65001
 
 # Russian code.page
 #~ code.page=1251 - Устанавливать НЕ рекомендуется!
-character.set=204
-output.code.page=1251
+character.set=0
+output.code.page=65001
 
 # OEM DOS866
 #~ code.page=866
@@ -570,10 +568,10 @@
 #~ output.code.page=65001
 
 # 1 включает перекодировку текста выводимого в окно консоли OEM в ANSI (только в SciTE-Ru)
-output.code.page.oem2ansi=1
+#output.code.page.oem2ansi=1
 
 # 1 включает детектирование кодировки dos866 (CodePage.lua)
-code.page.866.detect=1
+#code.page.866.detect=1
 
 # Автоматически включает кодировку UTF-8 при открытии UTF-8 файла не имеющего BOM заголовка (только в SciTE-Ru)
 # 1 - включает детектирование и автоматически добавляет BOM заголовок к редактируемому тексту,
@@ -712,6 +710,9 @@
 Alt+R|IDM_READONLY|\
 Ctrl+Shift+PageUp|IDM_MOVETABLEFT|\
 Ctrl+Shift+PageDown|IDM_MOVETABRIGHT|\
+Alt+Q|IDM_QUIT|\
+Ctrl+F1|IDM_HELP_SCITE|\
+Ctrl+Alt+F|IDM_TOGGLE_FOLDALL|\
 $(user.shortcuts.add)
 
 [ToolBar]            Настройка панели инструментов используемая по умолчанию
@@ -766,80 +767,55 @@
 #~ import toolbar\toolbar_gnome.properties
 import toolbar\toolbar_cool.properties
 
-[Styles]             Шрифты и цветовые настройки
-# Задание пользовательских переменных для шрифтов, используемых в стандартных стилях.
-# -----------------------------------------------------------
-# Переключение между наборами производится комбинацией Ctrl+F11 с помощью скрипта FontChanger.lua
-# Первая подборка шрифтов (фиксированный)
-font.0.base=font:Courier New,size:10
-font.0.small=font:Courier New,size:8
-font.0.comment=font:Courier New,size:9,italics
+[Styles]
+# Give symbolic names to the set of fonts used in the standard styles.
+font.base=font:Verdana,size:10
+font.small=font:Verdana,size:8
+font.comment=font:Comic Sans MS,size:9
+font.code.comment.box=$(font.comment)
+font.code.comment.line=$(font.comment)
+font.code.comment.doc=$(font.comment)
+font.code.comment.nested=$(font.comment)
+font.text=font:Times New Roman,size:11
+font.text.comment=font:Verdana,size:9
+font.embedded.base=font:Verdana,size:9
+font.embedded.comment=font:Comic Sans MS,size:8
+font.monospace=font:Courier New,size:10
+font.vbs=font:Lucida Sans Unicode,size:10
+font.js=$(font.comment)
 
-# Вторая подборка шрифтов (пропорциональный)
-font.1.base=font:Verdana,size:10
-font.1.small=font:Microsoft Sans Serif,size:8
-font.1.comment=font:Comic Sans MS,size:8
+# Give symbolic names to the set of colours used in the standard styles.
+colour.code.comment.box=fore:#007F00
+colour.code.comment.line=fore:#007F00
+colour.code.comment.doc=fore:#3F703F
+colour.code.comment.nested=fore:#A0C0A0
+colour.text.comment=fore:#000000FF,back:#D0F0D0
+colour.other.comment=fore:#007F00
+colour.embedded.comment=back:#E0EEFF
+colour.embedded.js=back:#F0F0FF
+colour.notused=back:#FF0000
 
-# Третья подборка шрифтов (фиксированный)
-#~ font.2.base=font:Consolas,size:10
-#~ font.2.small=font:Consolas,size:8
-#~ font.2.comment=font:Consolas,size:8,italics
-
-# Определяем какой из наборов шрифтов будет использоваться по умолчанию
-font.base=$(font.0.base)
-font.small=$(font.0.small)
-font.comment=$(font.0.comment)
-
-# Переменная font.monospace часто используется в файлах языков
-# а также используется в скрипте SciTEHexEdit.lua
-font.monospace=$(font.base)
-# -----------------------------------------------------------
-
-# Зададим типовые цвета, чтобы затем их использовать во всех языковых .properties.
-colour.comment=fore:#008000
-colour.code.comment.doc=$(colour.comment),back:#FEFEFE
-colour.embedded.comment=back:#E0EEFF
-
-colour.number=fore:#AC00A9
-colour.keyword=fore:#0000DF
-colour.string=fore:#5E5EAE
-colour.char=$(colour.string)
-colour.operator=fore:#FF0000
+colour.number=fore:#007F7F
+colour.keyword=fore:#00007F
+colour.string=fore:#7F007F
+colour.char=fore:#7F007F
+colour.operator=fore:#000000
 colour.preproc=fore:#7F7F00
 colour.error=fore:#FFFF00,back:#FF0000
-colour.whitespace=fore:#808080
 
-# Возможные значения параметров для строки style:
-# font:<name>, size:<number>, fore:<colour>, back:<colour>, italics|notitalics, bold|notbold, eolfilled|noteolfilled, underlined|notunderlined, case:m|u|l, visible|notvisible, changeable|notchangeable
-# Если строке стиля параметр упоминается дважды (следует избегать!), то действует крайнее правое значение.
-
-# Глобальные стили по умолчанию для всех языков программирования
-
-# Заданный по умолчанию стиль, и его особенности будут унаследованы всеми другими стилями
+# Global default styles for all languages
+# Default
 style.*.32=$(font.base)
-
-# Цвет фона номеров строк в режиме только для чтения (используется скриптом ReadOnly.lua)
-# Если не задан, то фон в режиме ReadOnly не меняется
-style.back.readonly=#FFEEEE
-
-# Стиль для номеров строк (с левого края)
-style.*.33=font:Microsoft Sans Serif,size:8,notbold
-
-# Стиль используется, чтобы отобразить парные скобки () или {}
-style.*.34=back:#99FFCC,bold
-
-# Стиль используется, чтобы отобразить незакрытые скобки ( или {
-style.*.35=back:#FFAAFF,bold
-# Примечание: Для того, чтобы видеть фоновый цвет в style.*.34 и style.*.35, необходимо либо запретить caret.line.back либо использовать прозрачность (caret.line.back.alpha)
-
-# Стиль для непечатных символов. Это не полный стиль. И передний план, и фоновые цвета для этих символов определены соответствующим Lexer-ом, а не этим стилем
+# Line number
+style.*.33=back:#C0C0C0,$(font.base)
+# Brace highlight
+style.*.34=fore:#0000FF,bold
+# Brace incomplete highlight
+style.*.35=fore:#FF0000,bold
+# Control characters
 style.*.36=
-
 # Indentation guides
-style.*.37=fore:#C0C0C0
-
-# Стиль всплывающих подсказок по синтаксису (берутся из API файлов)
-style.*.38=font:MS Sans Serif,size:9
+style.*.37=fore:#C0C0C0,back:#FFFFFF
 
 # -------------------------------------------------------------
 # Колонка для фолдинга (свёртывание/развёртывание процедур)
@@ -931,15 +907,12 @@
 indic.style.31=#11DDFF,@80
 
 [Caret]              Курсор
-# Скорость мерцания курсора (в миллисекундах)
 caret.period=500
-
-# Мерцание во всех выделениях
-#~ caret.additional.blinks=0
-
-# Ширина курсора
-#~ caret.width=2
-
+#caret.fore=#FF0000
+#caret.additional.blinks=0
+caret.width=2
+caret.line.back=#B0E0E6
+caret.line.back.alpha=63
 caret.policy.xslop=1
 caret.policy.width=20
 caret.policy.xstrict=0
@@ -950,10 +923,6 @@
 caret.policy.ystrict=1
 caret.policy.yeven=1
 caret.policy.yjumps=0
-#~ visible.policy.strict=1
-#~ visible.policy.slop=1
-#~ visible.policy.lines=4
-#~ caret.sticky=1
 
 [Abbrev & API]       Параметры автодополнения и всплывающих подсказок
 # При 1 SciTE после ввода каждой буквы будет услужливо выводить подсказку с возможными вариантами дополнения.
@@ -987,161 +956,20 @@
 keyXML=Shift+F12
 
 menu.language=\
-Text|txt|$(keyText)|\
-#~ Ada|ads||\
-#~ ABAP|abap||\
-AutoIt3|au3||\
-AutoHotkey|ahk||\
-#~ Apache Confi&g|conf||\
-#~ Atmel AVR|avr||\
-Assembler|asm||\
-#~ ASN.1|asn1||\
-#~ Avenue|ave||\
-#~ Baan|bc||\
-Basic|bas||\
-&Batch|bat||\
-#~ Bullant|ant||\
-&C / C++|c||\
-#~ CMake|cmake||\
-C&#|cs||\
-#~ COBOL|cob||\
-#~ Csound|orc||\
-CSS|css||\
-#~ D|d||\
-&Difference|diff||\
-#~ &Eiffel|e||\
-#~ Erlang|erl||\
-&Errorlist|err||\
-#~ Euphoria|e||\
-#~ FlagShip|prg||\
-#~ Forth|forth||\
-#~ &Fortran|f90||\
-#~ Gap|g||\
-#~ Haskell|hs||\
-H&ypertext|html|$(keyHTML)|\
-#~ &InnoSetup|iss||\
-&Java|java||\
-Java&Script|js|$(keyJS)|\
-#~ &Kix|kix||\
-#~ Lisp|lisp||\
-#~ Lot|lot||\
-#~ Lout|lt||\
-Lu&a|lua||\
-#~ Matlab|m.matlab||\
-&Makefile|mak|$(keyMake)|\
-#~ MetaPost|mp||\
-#~ MMIXAL|mms||\
-#~ Modula-3|m3||\
-&nnCron crontab|tab||\
-NSIS|nsi||\
-#~ Objective Caml|ml||\
-#~ Octave|m.octave||\
-#~ Opal|impl||\
-Pascal|pas||\
-Pe&rl|pl||\
-P&HP|php||\
-#~ P&LSQL|spec||\
-#~ P&ostScript|ps||\
-#~ P&OV-Ray SDL|pov||\
-PowerShell|ps1||\
-#~ PowerPro|powerpro||\
-&Properties|properties||\
-Pytho&n|py||\
-#~ R|R||\
-Reso&urce|rc||\
-#~ Ruby|rb||\
-Shell|sh||\
-S&QL|sql||\
-#~ Specman|e||\
-#~ &TCL|tcl||\
-#~ TeX|tex||\
-#~ &txt2tags|t2t||\
-&VB|vb||\
-VBScr&ipt|vbs||\
-#~ Verilog|v||\
-#~ VHDL|vhd||\
-&XML|xml|$(keyXML)|\
-#~ YAML|yaml||
+MQL|mql||
 
 [Languages]          Импорт всех специализированных файлов настройки для конкретных языков программирования
 # Файлы, в конце имени которых стоит символ подчёркивания немного отличаются от распространяемых с официальным дистрибутивом
 # Закомментируйте неиспользуемые вами языки!
-#~ import languages\abap
-#~ import languages\ada
-#~ import languages\avr
-import languages\asm
-#~ import languages\asn1
-import languages\au3
-import languages\ahk
-#~ import languages\ave
-#~ import languages\baan
-#~ import languages\blitzbasic
-#~ import languages\bullant
-#~ import languages\caml
 import languages\conf
-import languages\cpp
-#~ import languages\cmake
-#~ import languages\d
-#~ import languages\csound
-import languages\css
-#~ import languages\eiffel
-#~ import languages\erlang
-#~ import languages\escript
-#~ import languages\euphoria
-#~ import languages\flagship
-#~ import languages\forth
-#~ import languages\fortran
-#~ import languages\freebasic
-import languages\html
-#~ import languages\inno
-#~ import languages\kix
-#~ import languages\latex
-#~ import languages\lisp
-#~ import languages\lot
-#~ import languages\lout
+#import languages\cpp
 import languages\lua
-# Раскомментируйте для включения LuaInspect и script_lua лексера!
-#~ import languages\script_lua
-#~ import languages\matlab
-#~ import languages\metapost
-#~ import languages\mmixal
-import languages\nncrontab
-import languages\nsis
-#~ import languages\opal
-import languages\pascal
-import languages\perl
-#~ import languages\pov
-import languages\powershell
-#~ import languages\ps
-#~ import languages\purebasic
-import languages\python
-#~ import languages\rebol
-import languages\registry
-#~ import languages\ruby
-#~ import languages\scriptol
-import languages\text
-#~ import languages\smalltalk
-#~ import languages\spice
-#~ import languages\specman
-import languages\sql
-#~ import languages\tcl
-#~ import languages\tex
-import languages\vb
-#~ import languages\verilog
-#~ import languages\vhdl
-#~ import languages\yaml
-# Другие неучтённые языки (*.properties;*.abbrev;*.ini;*.inf;*.reg;.txt;*.log;*.lst;*.doc;*.bat;*.cmd; и др.)
-import languages\others
+#import languages\others
+import languages\mql
 
 [Open]               Настройки файлового фильтра в диалогах открытия и сохранения файла
 source.files=All Source|\
-*.asm;*asp;*.au3;*.c;*.cc;*.cpp;*.cxx;*.cs;*.h;*.hh;*.hxx;*.hpp;\
-*.rc;*.rc2;*.dlg;*.def;\
-*.vb;*.vbs;*.bas;*.frm;*.cls;*.ctl;\
-*.java;*.js;*.py;*.pl;*.rb;*.cgi;*.lua;*.conf;\
-make*;*.mak;\
-*.properties;*.xml;*.iface;*.bat;*.e;\
-*.sql;*.php;*.inc;*.css;*.js;*.html;*.htm;*.hta|
+*.mqh;*.mq4|
 
 all.files=All Files (*.*)|*.*|
 
@@ -1150,69 +978,19 @@
 open.filter=\
 $(source.files)\
 $(all.files)\
-$(filter.abap)\
-$(filter.ada)\
 $(filter.conf)\
-$(filter.ahk)\
-$(filter.asm)\
-$(filter.asn1)\
-$(filter.ave)\
-$(filter.baan)\
-$(filter.bash)\
-$(filter.caml)\
-$(filter.cmake)\
 $(filter.cpp)\
-#~ $(filter.ch)\
-$(filter.css)\
-$(filter.d)\
-$(filter.eiffel)\
-$(filter.erlang)\
-$(filter.fortran)\
-$(filter.gap)\
-#~ $(filter.hs)\
-$(filter.idl)\
-$(filter.inno)\
-$(filter.java)\
-$(filter.js)\
-$(filter.kix)\
-$(filter.lout)\
 $(filter.lua)\
-$(filter.matlab)\
-$(filter.metapost)\
-$(filter.mmixal)\
-#~ $(filter.modula3)\
-$(filter.nncrontab)\
-$(filter.nsis)\
-$(filter.opal)\
-$(filter.pascal)\
-$(filter.perl)\
-$(filter.php)\
-$(filter.pov)\
-$(filter.powershell)\
-$(filter.prg)\
 $(filter.properties)\
-$(filter.ps)\
-$(filter.python)\
-$(filter.r)\
-$(filter.ruby)\
-$(filter.sql)\
-$(filter.specman)\
-$(filter.tcl)\
-$(filter.tex)\
 $(filter.text)\
-$(filter.txt2tags)\
-$(filter.vb)\
-$(filter.web)\
-$(filter.yaml)\
-$(filter.verilog)\
-$(filter.vhdl)
+$(filter.mql)
 
 save.filter=$(open.filter)
 
 [LUA]                Подключение при загрузке внешнего LUA скрипта
 ext.lua.startup.script=$(SciteDefaultHome)\tools\SciTEStartup.lua
-#~ ext.lua.auto.reload=0
-#~ ext.lua.reset=1
+ext.lua.auto.reload=1
+ext.lua.reset=0
 
 [Macros]             Макросы
 # Путь до файла макросов (по умолчанию $(scite.userhome)\SciTE.macro или $(SciteDefaultHome)\SciTE.macro)
@@ -1236,7 +1014,7 @@
 
 [SideBar]            Настройки многофункциональной боковой панели (SideBar.lua)
 # Использовать для боковой панели отдельное окно: 1-да, 0-нет(default)
-#~ sidebar.win=1
+#sidebar.win=0
 # Показывать SideBar при запуске редактора: 1-да, 0-нет(default)
 sidebar.show=1
 # Позиция и ширина панели
@@ -1293,5 +1071,9 @@
 # Сохранять в файле сессии некоторые установки SciTE (save_settings.lua)
 save.settings=1
 
+#import themes\dark
+#import themes\vim
+#import themes\solarized
+
 # Делаем доступными переменные из файла SciTE.session
 import home\SciTE.session
diff -r 57c52893ca01 -r 7bdba7c9c807 pack/tools/COMMON.lua
--- a/pack/tools/COMMON.lua	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/tools/COMMON.lua	Wed Apr 22 22:14:56 2015 -0500
@@ -264,6 +264,65 @@
 		end)
 	end)
 end
+------------------------------------------
+-- ifnil
+------------------------------------------
+function ifpnil(prop, def)
+	local val = props[prop]
+	if val == nil or val == '' then
+		return def
+	else
+		return val
+	end
+end
+------------------------------------------
+-- Translate color from RGB to win
+------------------------------------------
+local function encodeRGB2WIN(color)
+	if color == nil then return nil end
+	if string.sub(color,1,1)=="#" and string.len(color)>6 then
+		return tonumber(string.sub(color,6,7)..string.sub(color,4,5)..string.sub(color,2,3), 16)
+	else
+		return color
+	end
+end
+------------------------------------------
+-- Translate INDIC_*
+------------------------------------------
+local function GetStyle(mark_string)
+	local mark_style_table = {
+		plain    = INDIC_PLAIN,    squiggle = INDIC_SQUIGGLE,
+		tt       = INDIC_TT,       diagonal = INDIC_DIAGONAL,
+		strike   = INDIC_STRIKE,   hidden   = INDIC_HIDDEN,
+		roundbox = INDIC_ROUNDBOX, box      = INDIC_BOX
+	}
+	return mark_style_table[mark_string]
+end
+------------------------------------------
+-- Init Mark Style
+------------------------------------------
+local function InitMarkStyle(mark_number, mark_style, color, alpha)
+	editor.IndicStyle[mark_number] = mark_style
+	editor.IndicFore[mark_number]  = encodeRGB2WIN(color)
+	editor.IndicAlpha[mark_number] = alpha
+	editor.IndicUnder[mark_number] = true
+end
+------------------------------------------
+-- Parse Mark Style From Prop String
+------------------------------------------
+function ParseMarkStyle(prop_string)
+	local mark = props[prop_string]
+	local ret_number = 8
+	if mark ~= "" then
+		local mark_number= tonumber(mark:match("%d+")) or 30
+		local mark_color = mark:match("#%x%x%x%x%x%x") or (props["find.mark"]):match("#%x%x%x%x%x%x") or "#0F0F0F"
+		local mark_style = GetStyle(mark:match("%l+")) or INDIC_ROUNDBOX
+		local alpha_fill = tonumber((mark:match("%@%d+") or ""):sub(2)) or 30
+		InitMarkStyle(mark_number, mark_style, mark_color, alpha_fill)
+		ret_number = mark_number
+	end
+	return ret_number
+end
 
 -- ==============================================================
 --- Читает файлы .abbrev (понимает инструкцию #import)
diff -r 57c52893ca01 -r 7bdba7c9c807 pack/tools/CodePage.lua
--- a/pack/tools/CodePage.lua	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/tools/CodePage.lua	Wed Apr 22 22:14:56 2015 -0500
@@ -46,7 +46,7 @@
 		elseif props["character.set"]=='204' then
 			props["editor.code.page.name"]='WIN-1251'
 		elseif tonumber(props["character.set"])==0 then
-			props["editor.code.page.name"]='CP1252'
+			props["editor.code.page.name"]='DEFAULT'
 		elseif props["character.set"]=='238' then
 			props["editor.code.page.name"]='CP1250'
 		elseif props["character.set"]=='161' then
diff -r 57c52893ca01 -r 7bdba7c9c807 pack/tools/SciTEStartup.lua
--- a/pack/tools/SciTEStartup.lua	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/tools/SciTEStartup.lua	Wed Apr 22 22:14:56 2015 -0500
@@ -98,7 +98,8 @@
 dofile (props["SciteDefaultHome"].."\\tools\\ShowCalltip.lua")
 
 -- Авто подсветка текста, который совпадает с текущим словом или выделением
-dofile (props["SciteDefaultHome"].."\\tools\\highlighting_identical_text.lua")
+--dofile (props["SciteDefaultHome"].."\\tools\\highlighting_identical_text.lua")
+dofile (props["SciteDefaultHome"].."\\tools\\SmartHighlight.lua")
 
 -- Подсветка, копирование, вставка, удаление парных тегов в HTML
 dofile (props["SciteDefaultHome"].."\\tools\\paired_tags.lua")
diff -r 57c52893ca01 -r 7bdba7c9c807 pack/tools/SideBar.lua
--- a/pack/tools/SideBar.lua	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/tools/SideBar.lua	Wed Apr 22 22:14:56 2015 -0500
@@ -208,10 +208,10 @@
 -- tab0:memo_path   Path and Mask
 ----------------------------------------------------------
 local current_path = ''
-local file_mask = '*.*'
+local file_mask = '*.mq4'
 
 local function FileMan_ShowPath()
-	local rtf = [[{\rtf{\fonttbl{\f0\fcharset204 Helv;}}{\colortbl;\red0\green0\blue255;\red255\green0\blue0;}\f0\fs16]]
+	local rtf = ifpnil('sidebar.filepath.style', [[{\rtf{\fonttbl{\f0\fcharset204 Helv;}}{\colortbl;\red0\green0\blue255;\red255\green0\blue0;}\f0\fs16]])
 	local path = '\\cf1'..current_path:gsub('\\', '\\\\')
 	local mask = '\\cf2'..file_mask..'}'
 	memo_path:set_text(rtf..path..mask)
@@ -584,6 +584,7 @@
 -- 3 - function paramaters with parentheses
 local _sort = 'order'
 local _backjumppos -- store position if jumping
+local functions_filename = props['SciteDefaultHome']..'\\api\\dynamic.api'
 
 local Lang2lpeg = {}
 do
@@ -1075,6 +1076,7 @@
 	-- lpegPattern = nil
 	table_functions = lpegPattern:match(textAll, start_code_pos+1) -- 2nd arg is the symbol index to start with
 	_DEBUG.timerstop('Functions_GetNames','lpeg')
+	Functions_SaveList()
 end
 
 local function Functions_ListFILL()
@@ -1125,6 +1127,48 @@
 	end
 end
 
+function Functions_SaveList()
+    --print("saving ",functions_filename)
+	local function emptystr(...) return '' end
+	
+	local function GetParams (funcitem)
+		return (funcitem[3] and ' '..funcitem[3]) or ''
+	end
+	
+	local function GetFlags (funcitem)
+		local res = ''
+		local add = ''
+		for flag,value in pairs(funcitem) do
+			if type(flag)=='string' then
+				if type(value)=='string' then	add = flag .. value
+				elseif type(value)=='number' then add = flag..':'..value
+				else add = flag end
+				res = res .. '['.. add ..']'
+			end
+		end
+		if res~='' then res = res .. ' ' end
+		return res or ''
+	end
+
+	if not _show_flags then GetFlags = emptystr end
+
+	local function fixname (funcitem)
+		return funcitem[1]..GetParams(funcitem)
+	end
+	
+	if pcall(io.output, functions_filename) then
+		for _, a in ipairs(table_functions) do
+			local funcname = fixname(a)
+			if tonumber(props["editor.unicode.mode"]) == IDM_ENCODING_DEFAULT then
+				funcname = funcname:to_utf8(editor:codepage())
+			end
+			io.write(funcname.."\n")
+		end
+		io.close()
+    scite.Perform('reloadproperties:')
+	end
+end
+
 function Functions_SortByOrder()
 	_sort = 'order'
 	Functions_ListFILL()
@@ -1151,6 +1195,8 @@
 	local pos = list_func:get_item_data(sel_item)
 	if pos then
 		ShowCompactedLine(pos)
+		editor:GotoLine(pos + editor.LinesOnScreen/3*2)
+		if editor.FirstVisibleLine > (pos - 10) then editor:GotoLine(pos - editor.LinesOnScreen/3) end
 		editor:GotoLine(pos)
 		gui.pass_focus()
 	end
diff -r 57c52893ca01 -r 7bdba7c9c807 pack/tools/lexer_name.lua
--- a/pack/tools/lexer_name.lua	Mon Apr 20 04:26:53 2015 +0300
+++ b/pack/tools/lexer_name.lua	Wed Apr 22 22:14:56 2015 -0500
@@ -19,6 +19,8 @@
 	if cur_lexer ~= last_lexer then
 		if cur_lexer == "hypertext" then
 			props["scite.lexer.name"] = "html"
+    elseif props['FileExt'] == 'mq4' or props['FileExt'] == 'mqh' then
+      props['scite.lexer.name'] = 'mql'
 		else
 			props["scite.lexer.name"] = cur_lexer
 		end
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scintilla/include/SciLexer.h
--- a/src/scintilla/include/SciLexer.h	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scintilla/include/SciLexer.h	Wed Apr 22 22:14:56 2015 -0500
@@ -180,6 +180,10 @@
 #define SCE_C_USERLITERAL 25
 #define SCE_C_TASKMARKER 26
 #define SCE_C_ESCAPESEQUENCE 27
+#define SCE_C_KEYWORD6 28
+#define SCE_C_KEYWORD7 29
+#define SCE_C_KEYWORD8 30
+#define SCE_C_KEYWORD9 31
 #define SCE_D_DEFAULT 0
 #define SCE_D_COMMENT 1
 #define SCE_D_COMMENTLINE 2
@@ -506,6 +510,8 @@
 #define SCE_ERR_JAVA_STACK 20
 #define SCE_ERR_VALUE 21
 #define SCE_ERR_GCC_INCLUDED_FROM 22
+#define SCE_ERR_MQL 23
+#define SCE_ERR_MQLWRN   24
 #define SCE_ERR_FIND_VALUE 30
 #define SCE_BAT_DEFAULT 0
 #define SCE_BAT_COMMENT 1
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scintilla/include/Scintilla.iface
--- a/src/scintilla/include/Scintilla.iface	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scintilla/include/Scintilla.iface	Wed Apr 22 22:14:56 2015 -0500
@@ -2842,6 +2842,10 @@
 val SCE_C_USERLITERAL=25
 val SCE_C_TASKMARKER=26
 val SCE_C_ESCAPESEQUENCE=27
+val SCE_C_KEYWORD6=28
+val SCE_C_KEYWORD7=29
+val SCE_C_KEYWORD8=30
+val SCE_C_KEYWORD9=31
 # Lexical states for SCLEX_D
 lex D=SCLEX_D SCE_D_
 val SCE_D_DEFAULT=0
@@ -3207,6 +3211,8 @@
 val SCE_ERR_JAVA_STACK=20
 val SCE_ERR_VALUE=21
 val SCE_ERR_GCC_INCLUDED_FROM=22
+val SCE_ERR_MQL=23
+val SCE_ERR_MQLWRN=24
 ##!-start-[FindResultListStyle]
 val SCE_ERR_FIND_VALUE=30
 ##!-end-[FindResultListStyle]
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scintilla/lexers/LexCPP.cxx
--- a/src/scintilla/lexers/LexCPP.cxx	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scintilla/lexers/LexCPP.cxx	Wed Apr 22 22:14:56 2015 -0500
@@ -353,6 +353,10 @@
             "Global classes and typedefs",
             "Preprocessor definitions",
             "Task marker and error marker keywords",
+			"Third utility keywords set",
+			"Forth utility keywords set",
+			"Fifth utility keywords set",
+			"Sixth utility keywords set",
             0,
 };
 
@@ -445,6 +449,11 @@
 	WordList keywords4;
 	WordList ppDefinitions;
 	WordList markerList;
+	WordList keywords6;
+	WordList keywords7;
+	WordList keywords8;
+	WordList keywords9;
+
 	struct SymbolValue {
 		std::string value;
 		std::string arguments;
@@ -589,6 +598,18 @@
 	case 5:
 		wordListN = &markerList;
 		break;
+	case 6:
+		wordListN = &keywords6;
+		break;
+	case 7:
+		wordListN = &keywords7;
+		break;
+	case 8:
+		wordListN = &keywords8;
+		break;
+	case 9:
+		wordListN = &keywords9;
+		break;
 	}
 	int firstModification = -1;
 	if (wordListN) {
@@ -810,6 +831,14 @@
 						sc.ChangeState(SCE_C_WORD2|activitySet);
 					} else if (keywords4.InList(s)) {
 						sc.ChangeState(SCE_C_GLOBALCLASS|activitySet);
+					} else if (keywords6.InList(s)) {
+						sc.ChangeState(SCE_C_KEYWORD6|activitySet);
+					} else if (keywords7.InList(s)) {
+						sc.ChangeState(SCE_C_KEYWORD7|activitySet);
+					} else if (keywords8.InList(s)) {
+						sc.ChangeState(SCE_C_KEYWORD8|activitySet);
+					} else if (keywords9.InList(s)) {
+						sc.ChangeState(SCE_C_KEYWORD9|activitySet);
 					} else {
 						int subStyle = classifierIdentifiers.ValueFor(s);
 						if (subStyle >= 0) {
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scintilla/lexers/LexOthers.cxx
--- a/src/scintilla/lexers/LexOthers.cxx	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scintilla/lexers/LexOthers.cxx	Wed Apr 22 22:14:56 2015 -0500
@@ -1244,6 +1244,7 @@
 		// CTags: <identifier>\t<filename>\t<message>
 		// Lua 5 traceback: \t<filename>:<line>:<message>
 		// Lua 5.1: <exe>: <filename>:<line>:<message>
+		// MQL: <filename>(<line>,<column>) : error|warning <num>: <message>
 		bool initialTab = (lineBuffer[0] == '\t');
 		bool initialColonPart = false;
 		bool canBeCtags = !initialTab;	// For ctags must have an identifier with no spaces then a tab
@@ -1251,7 +1252,8 @@
 			stGccStart, stGccDigit, stGccColumn, stGcc,
 			stMsStart, stMsDigit, stMsBracket, stMsVc, stMsDigitComma, stMsDotNet,
 			stCtagsStart, stCtagsFile, stCtagsStartString, stCtagsStringDollar, stCtags,
-			stUnrecognized
+			stUnrecognized,
+			stMql
 		} state = stInitial;
 		for (unsigned int i = 0; i < lengthLine; i++) {
 			char ch = lineBuffer[i];
@@ -1333,7 +1335,6 @@
 			} else if (state == stMsDigitComma) {	// <filename>(<line>,
 				if (ch == ')') {
 					state = stMsDotNet;
-					break;
 				} else if ((ch != ' ') && !Is0To9(ch)) {
 					state = stUnrecognized;
 				}
@@ -1352,6 +1353,12 @@
 			} else if ((state == stCtagsStartString) && ((lineBuffer[i] == '$') && (lineBuffer[i + 1] == '/'))) {
 				state = stCtagsStringDollar;
 				break;
+			} else if (state == stMsDotNet) {
+				if (ch == ' ' && chNext == ':' && lineBuffer[i + 2] == ' ') 
+					state = stMql;
+			} else if (state == stMql) {
+				if (ch == 'e') return SCE_ERR_MQL;
+				if (ch == 'w') return SCE_ERR_MQLWRN;
 			}
 		}
 		if (state == stGcc) {
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scite/src/IFaceTable.cxx
--- a/src/scite/src/IFaceTable.cxx	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scite/src/IFaceTable.cxx	Wed Apr 22 22:14:56 2015 -0500
@@ -744,6 +744,10 @@
 	{"SCE_C_VERBATIM",13},
 	{"SCE_C_WORD",5},
 	{"SCE_C_WORD2",16},
+	{"SCE_C_KEYWORD6",28},
+	{"SCE_C_KEYWORD7",29},
+	{"SCE_C_KEYWORD8",30},
+	{"SCE_C_KEYWORD9",31},
 	{"SCE_DIFF_ADDED",6},
 	{"SCE_DIFF_CHANGED",7},
 	{"SCE_DIFF_COMMAND",2},
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scite/src/SciTEBuffers.cxx
--- a/src/scite/src/SciTEBuffers.cxx	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scite/src/SciTEBuffers.cxx	Wed Apr 22 22:14:56 2015 -0500
@@ -1959,7 +1959,32 @@
 			}
 			break;
 		}
+		
+	case SCE_ERR_MQLWRN:
+	case SCE_ERR_MQL: {
+			const char *start = cdoc;
+			while (isspacechar(*start)) {
+				start++;
+			}
+			const char *endPath = strchr(start, '(');
+			if (endPath) {
+				if (!isdigitchar(endPath[1])) {
+					// This handles the common case of include files in the C:\Program Files (x86)\ directory
+					endPath = strchr(endPath + 1, '(');
+				}
+				if (endPath) {
+					ptrdiff_t length = endPath - start;
+					sourcePath.assign(start, length);
+					endPath++;
+					const char *startCol = strchr(endPath, ',') + 1;
+					column = atoi(startCol);
+					return atoi(endPath) - 1;
+				}
+			}
+			break;
+		}
 	}	// switch
+
 	return -1;
 }
 
diff -r 57c52893ca01 -r 7bdba7c9c807 src/scite/src/SciTEProps.cxx
--- a/src/scite/src/SciTEProps.cxx	Mon Apr 20 04:26:53 2015 +0300
+++ b/src/scite/src/SciTEProps.cxx	Wed Apr 22 22:14:56 2015 -0500
@@ -664,7 +664,8 @@
 		ForwardPropertyToEditor(propertiesToForward[i]);
 	}
 
-	if (apisFileNames != props.GetNewExpandString("api.", fileNameForExtension.c_str())) {
+	//if (apisFileNames != props.GetNewExpandString("api.", fileNameForExtension.c_str())) {
+  {
 		apis.Clear();
 		ReadAPI(fileNameForExtension);
 		apisFileNames = props.GetNewExpandString("api.", fileNameForExtension.c_str());
